crypto/ec/ecp_nistp384.c

```diff
@@ -0,0 +1,1935 @@
+static void bin48_to_felem(felem out, const u8 in[48])
+{
+    memset(out, 0, 56);
+    out[0] = (*((limb *) & in[0])) & bottom56bits;
+    out[1] = (*((limb_aX *) & in[7])) & bottom56bits;
+    out[2] = (*((limb_aX *) & in[14])) & bottom56bits;
+    out[3] = (*((limb_aX *) & in[21])) & bottom56bits;
+    out[4] = (*((limb_aX *) & in[28])) & bottom56bits;
+    out[5] = (*((limb_aX *) & in[35])) & bottom56bits;
+    memmove(&out[6], &in[42], 6);
+}
+
+static void felem_to_bin48(u8 out[48], const felem in)
+{
+    memset(out, 0, 48);
+    (*((limb *) & out[0]))     |= (in[0] & bottom56bits);
+    (*((limb_aX *) & out[7]))  |= (in[1] & bottom56bits);
+    (*((limb_aX *) & out[14])) |= (in[2] & bottom56bits);
+    (*((limb_aX *) & out[21])) |= (in[3] & bottom56bits);
+    (*((limb_aX *) & out[28])) |= (in[4] & bottom56bits);
+    (*((limb_aX *) & out[35])) |= (in[5] & bottom56bits);
+    memmove(&out[42], &in[6], 6);
+}
+
+/* BN_to_felem converts an OpenSSL BIGNUM into an felem */
+static int BN_to_felem(felem out, const BIGNUM *bn)
+{
+    felem_bytearray b_out;
+    int num_bytes;
+
+    if (BN_is_negative(bn)) {
+        ERR_raise(ERR_LIB_EC, EC_R_BIGNUM_OUT_OF_RANGE);
+        return 0;
+    }
+    num_bytes = BN_bn2lebinpad(bn, b_out, sizeof(b_out));
+    if (num_bytes < 0) {
+        ERR_raise(ERR_LIB_EC, EC_R_BIGNUM_OUT_OF_RANGE);
+        return 0;
+    }
+    bin48_to_felem(out, b_out);
+    return 1;
+}
+
+/* felem_to_BN converts an felem into an OpenSSL BIGNUM */
+static BIGNUM *felem_to_BN(BIGNUM *out, const felem in)
+{
+    felem_bytearray b_out;
+    felem_to_bin48(b_out, in);
+    return BN_lebin2bn(b_out, sizeof(b_out), out);
+}
+
+/*-
+ * Field operations
+ * ----------------
+ */
+
+static void felem_one(felem out)
+{
+    unsigned i;
+    out[0] = 1;
+    for (i = 1; i < NLIMBS; i++)
+        out[i] = 0;
+}
+
+static void felem_assign(felem out, const felem in)
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++)
+        out[i] = in[i];
+}
+
+/* felem_sum64 sets out = out + in. */
+static void felem_sum64(felem out, const felem in)
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++)
+        out[i] += in[i];
+}
+
+/* felem_scalar sets out = in * scalar */
+static void felem_scalar(felem out, const felem in, limb scalar)
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++)
+        out[i] = in[i] * scalar;
+}
+
+/* felem_scalar64 sets out = out * scalar */
+static void felem_scalar64(felem out, limb scalar)
+{
+    unsigned i;
+    for (i = 0; i < NLIMBS; i++)
+        out[i] *= scalar;
+}
+
+/* felem_scalar128 sets out = out * scalar */
+static void felem_scalar128(widefelem out, limb scalar)
+{
+    unsigned i;
+    for (i = 0; i < 2*NLIMBS-1; i++)
+        out[i] *= scalar;
+}
+
+/*-
+ * felem_neg sets |out| to |-in|
+ * On entry:
+ *   in[i] < 2^60 - 2^29
+ * On exit:
+ *   out[i] < 2^60
+ */
+static void felem_neg(felem out, const felem in)
+{
+    /* In order to prevent underflow, we subtract from 0 mod p. */
```

Comment:
More explanation would be nice.  The subtraction doesn't look to be from zero.
_subtract from p which is 0 mod p_ perhaps?

Reply:
I meant subtract from the congruence class containing zero. Let's change that to 'subtract from a multiple of p'.